### PR TITLE
Implement CampaignSpec 

### PR DESCRIPTION
**Pull request by uakyol**
This PR implements the mechanism to generate impression data for a given campaign for an EDP.

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+DAILY_NOISE_FACTOR = 0.1
+
+# Acceptable change applied to max frequency to make the num impressions
+# play well with freq distribution.
+ACCEPTABLE_FREQ_DIST_CORRECTION_FACTOR = 0.1
+
+
+class CampaignSpec:
+    """Class that samples impressions given specification for a single campaign.
+
```
Comment (jiayupeng-ds) [inline generate.py:42]: Before going to details, you may summarzie the functionality of this class as "Samples impressions on a given EDP on given dates, such that they approximately align with the given number of impressions, reach, and distributions of frequency, video completion, and viewability."

File: generate.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    1. At initialization normalizes and reconstructs the frequency distribution to pad the frequency to meet the requirements of the other parameters.
+        E.g. If freq distrubution is specified by [(1, 800), (2, 600), (3, 500)] 800 impressions with freq=1, 600 with freq=2...
+             Then converts it to the prob distribution [(1, 0.42), (2, 0.31), (3, 0.27)]
+             Given freq dist implies 1*800+ 2*600 + 3*500 = 4100 impressions but the given impresison requirement can be different (e.g. 5000)
+                If so, then pads another frequency (4) to match that 4100+ 4*225 = 5000.
+             Also makes sure that this correction does not change the given freq dist more than a small amount
+    2. At initialization creates a pool of vids to be used in the sampling. This pool of vids consists of repeats specified by the frequency distribution.
+        E.g : If the freq dist specifies 5 impressions with freq=1 and 3 with freq=2 using vidSet {1..100} we can generate
+              vids=[1, 2, 3, 4, 5, 6 , 6, 7, 7, 8, 8]
```
Comment (jiayupeng-ds) [inline generate.py]: Add: Randomly shuffle the pool of vids.

File: generate.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    3. Selects number of impressions for each day = (total_impressions/numdays) + noise.  Where noise is uniform.
+    3. For each day, pops vids from the pool equal to the number of impressions for that day.
```
Comment (jiayupeng-ds) [inline generate.py]: equal to -> according to?

File: generate_test.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        )
+
+        generated_number_of_impressions = len(impressions)
+        self.assertTrue(
+            generated_number_of_impressions,
+            approx(total_impressions, rel=ACCEPTABLE_RELATIVE_ERROR),
+        )
+
+        generated_reach = len(set(list(map(lambda x: x.vid, impressions))))
+        self.assertTrue(generated_reach, approx(total_reach, rel=ACCEPTABLE_RELATIVE_ERROR))
```
Comment (jiayupeng-ds) [inline generate_test.py]: May also test the alignment with the distributions of frequency, completion, and viewability

**Review by jiayu-google — APPROVED** [review]
LGTM as a one-off script to generate synthetic data.

In the future if you want to repeatedly use this script, you may come back and polish it, for example, add typings and docstrings.

File: generate.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    4. For each impression, samples the video completion and viewability specified by the given distributions for them.
```
Comment (jiayupeng-ds) [inline generate.py]: nit: samples -> independently samples

**Review by ghost — APPROVED** [review]

**Review by zasadnyy — APPROVED** [review]
Looks good, thanks @uakyol

File: generate.py (hunk 2 above)
Comment (uakyol) [reply generate.py]: thanks

File: generate.py (hunk 1 above)
Comment (uakyol) [reply generate.py:42]: awesome, thank you

File: generate.py (hunk 5 above)
Comment (uakyol) [reply generate.py]: yep, thank you

File: generate.py (hunk 3 above)
Comment (uakyol) [reply generate.py]: sure, thank you

File: generate_test.py (hunk 4 above)
Comment (uakyol) [reply generate_test.py]: Thanks a lot for pointing out Jiayu, I added them. Also, while adding I realized I was asserting incorrectly. I corrected it and increases the acceptable relative error to 0.1. 
Can you please have a look and let me know if this looks good to you?

**Review by mariolamassaavedra — APPROVED** [review]

File: generate_test.py (hunk 4 above)
Comment (jiayupeng-ds) [reply generate_test.py]: Thank you for catching this, Ugur! (Good to learn pytest.approx) LGTM.

**Pull request merged**